File: kinesis/src/test/scala/akka/stream/alpakka/kinesis/scaladsl/Examples.scala
/*
 * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.stream.alpakka.kinesis.scaladsl

import java.nio.ByteBuffer
import java.util.Date

import akka.NotUsed
import akka.actor.ActorSystem
import akka.stream.alpakka.kinesis.{KinesisFlowSettings, ShardSettings}
import akka.stream.scaladsl.{Flow, Sink, Source}
import akka.stream.{ActorMaterializer, Materializer}
import akka.util.ByteString
import com.amazonaws.services.kinesis.model.{PutRecordsRequestEntry, PutRecordsResultEntry, Record, ShardIteratorType}
import com.amazonaws.services.kinesis.{AmazonKinesisAsync, AmazonKinesisAsyncClientBuilder}

import scala.concurrent.duration._

object Examples {

  //#init-client
  implicit val system: ActorSystem = ActorSystem()
  implicit val materializer: Materializer = ActorMaterializer()

  implicit val amazonKinesisAsync: com.amazonaws.services.kinesis.AmazonKinesisAsync =
    AmazonKinesisAsyncClientBuilder.defaultClient()

  system.registerOnTermination(amazonKinesisAsync.shutdown())
  //#init-client

  //#source-settings
  val settings = ShardSettings(streamName = "myStreamName",
                               shardId = "shard-id",
                               shardIteratorType = ShardIteratorType.TRIM_HORIZON,
                               refreshInterval = 1.second,
                               limit = 500)
  //#source-settings

  //#source-single
  val source: Source[com.amazonaws.services.kinesis.model.Record, NotUsed] =
    KinesisSource.basic(settings, amazonKinesisAsync)
  //#source-single

  //#source-list
  val mergeSettings = List(
    ShardSettings("myStreamName",
                  "shard-id-1",
                  ShardIteratorType.AT_SEQUENCE_NUMBER,
                  startingSequenceNumber = Some("sequence"),
                  refreshInterval = 1.second,
                  limit = 500),
    ShardSettings("myStreamName",
                  "shard-id-2",
                  ShardIteratorType.AT_TIMESTAMP,
                  atTimestamp = Some(new Date()),
                  refreshInterval = 1.second,
                  limit = 500)
  )

  val mergedSource: Source[Record, NotUsed] = KinesisSource.basicMerge(mergeSettings, amazonKinesisAsync)
  //#source-list

  //#flow-settings
  val flowSettings = KinesisFlowSettings(
    parallelism = 1,
    maxBatchSize = 500,
    maxRecordsPerSecond = 1000,
    maxBytesPerSecond = 1000000,
    maxRetries = 5,
    backoffStrategy = KinesisFlowSettings.Exponential,
    retryInitialTimeout = 100.millis
  )

  val defaultFlowSettings = KinesisFlowSettings.defaultInstance

  val fourShardFlowSettings = KinesisFlowSettings.byNumberOfShards(4)
  //#flow-settings

  //#flow-sink
  val flow1: Flow[PutRecordsRequestEntry, PutRecordsResultEntry, NotUsed] = KinesisFlow("myStreamName")

  val flow2: Flow[PutRecordsRequestEntry, PutRecordsResultEntry, NotUsed] = KinesisFlow("myStreamName", flowSettings)

  val flow3: Flow[(String, ByteString), PutRecordsResultEntry, NotUsed] =
    KinesisFlow.byPartitionAndBytes("myStreamName")

  val flow4: Flow[(String, ByteBuffer), PutRecordsResultEntry, NotUsed] =
    KinesisFlow.byPartitionAndData("myStreamName")

  val sink1: Sink[PutRecordsRequestEntry, NotUsed] = KinesisSink("myStreamName")
  val sink2: Sink[PutRecordsRequestEntry, NotUsed] = KinesisSink("myStreamName", flowSettings)
  val sink3: Sink[(String, ByteString), NotUsed] = KinesisSink.byPartitionAndBytes("myStreamName")
  val sink4: Sink[(String, ByteBuffer), NotUsed] = KinesisSink.byPartitionAndData("myStreamName")
  //#flow-sink

}

File: kinesis/src/test/java/akka/stream/alpakka/kinesis/javadsl/Examples.java
/*
 * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.stream.alpakka.kinesis.javadsl;

import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.japi.Pair;
import akka.stream.ActorMaterializer;
import akka.stream.alpakka.kinesis.KinesisFlowSettings;
import akka.stream.alpakka.kinesis.ShardSettings;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import com.amazonaws.services.kinesis.AmazonKinesisAsync;
import com.amazonaws.services.kinesis.AmazonKinesisAsyncClientBuilder;
import com.amazonaws.services.kinesis.model.PutRecordsRequestEntry;
import com.amazonaws.services.kinesis.model.PutRecordsResultEntry;
import com.amazonaws.services.kinesis.model.Record;
import com.amazonaws.services.kinesis.model.ShardIteratorType;
import scala.concurrent.duration.FiniteDuration;

import java.util.Arrays;
import java.util.concurrent.TimeUnit;

public class Examples {

  // #init-client
  final ActorSystem system = ActorSystem.create();
  final ActorMaterializer materializer = ActorMaterializer.create(system);

  final com.amazonaws.services.kinesis.AmazonKinesisAsync amazonKinesisAsync =
      AmazonKinesisAsyncClientBuilder.defaultClient();
  // #init-client

  {
    // #init-client

    system.registerOnTermination(amazonKinesisAsync::shutdown);
    // #init-client
  }

  // #source-settings
  final ShardSettings settings =
      ShardSettings.create("streamName", "shard-id")
          .withShardIteratorType(ShardIteratorType.LATEST)
          .withRefreshInterval(1L, TimeUnit.SECONDS)
          .withLimit(500);
  // #source-settings

  // #source-single
  final Source<com.amazonaws.services.kinesis.model.Record, NotUsed> source =
      KinesisSource.basic(settings, amazonKinesisAsync);
  // #source-single

  // #source-list
  final Source<Record, NotUsed> two =
      KinesisSource.basicMerge(Arrays.asList(settings), amazonKinesisAsync);
  // #source-list

  // #flow-settings
  final KinesisFlowSettings flowSettings =
      KinesisFlowSettings.create()
          .withParallelism(1)
          .withMaxBatchSize(500)
          .withMaxRecordsPerSecond(1_000)
          .withMaxBytesPerSecond(1_000_000)
          .withMaxRecordsPerSecond(5)
          .withBackoffStrategyExponential()
          .withRetryInitialTimeout(100L, TimeUnit.MILLISECONDS);

  final KinesisFlowSettings defaultFlowSettings = KinesisFlowSettings.create();

  final KinesisFlowSettings fourShardFlowSettings = KinesisFlowSettings.byNumberOfShards(4);
  // #flow-settings

  // #flow-sink
  final Flow<PutRecordsRequestEntry, PutRecordsResultEntry, NotUsed> flow =
      KinesisFlow.apply("streamName", flowSettings, amazonKinesisAsync);

  final Flow<PutRecordsRequestEntry, PutRecordsResultEntry, NotUsed> defaultSettingsFlow =
      KinesisFlow.apply("streamName", amazonKinesisAsync);

  final Flow<Pair<PutRecordsRequestEntry, String>, Pair<PutRecordsResultEntry, String>, NotUsed>
      flowWithStringContext =
          KinesisFlow.withUserContext("streamName", flowSettings, amazonKinesisAsync);

  final Flow<Pair<PutRecordsRequestEntry, String>, Pair<PutRecordsResultEntry, String>, NotUsed>
      defaultSettingsFlowWithStringContext =
          KinesisFlow.withUserContext("streamName", flowSettings, amazonKinesisAsync);

  final Sink<PutRecordsRequestEntry, NotUsed> sink =
      KinesisSink.apply("streamName", flowSettings, amazonKinesisAsync);

  final Sink<PutRecordsRequestEntry, NotUsed> defaultSettingsSink =
      KinesisSink.apply("streamName", amazonKinesisAsync);
  // #flow-sink

}
